### PR TITLE
fix(cleansoftware): incorrect max limit

### PR DESCRIPTION
**Pull request by flegastelois**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #11962

### CODE REVIEW

**Review by AdrienClairembault — APPROVED** [review]

File: src/CleanSoftwareCron.php [hunk 1]
```diff
@@ -93,7 +93,7 @@ public static function run(?int $max): int
         $total += self::deleteItems(
             self::getSoftwareWithNoVersionsCriteria(),
             new Software(),
-            $max - $total
+            $max
```
Comment (cedric-anne) [inline src/CleanSoftwareCron.php:96]: Why did you remove this ?
If call on line 86 already removed the X elements, then this call should not remove more than `$max - X` element.

File: src/CleanSoftwareCron.php [hunk 2]
```diff
@@ -191,21 +191,13 @@ protected static function deleteItems(
...
         global $DB;
 
-        $total = 0;
-
-        do {
-            $scope['LIMIT'] = min($max - $total, self::MAX_BATCH_SIZE);
-            $items = $DB->request($scope);
-            $count = count($items);
-            $total += $count;
+        $scope['LIMIT'] = min($max, self::MAX_BATCH_SIZE);
```
Comment (cedric-anne) [inline src/CleanSoftwareCron.php:194]: `MAX_BATCH_SIZE` is made to limit query results, not to limit maximum items to delete.
For instance, you may want to delete 10000 items, and, in this case, it will do it in 5 batches of 2000 items.

File: src/CleanSoftwareCron.php (hunk 1 above)
Comment (flegastelois) [reply src/CleanSoftwareCron.php:96]: Functionally speaking, we want to delete a maximum number of version (ex: 200) and a maximum number of software (ex: 200). If I indicate "200", it must apply as is for both requests. Without this change, there might even be cases where software are never deleted.

Comment (AdrienClairembault) [reply src/CleanSoftwareCron.php:96]: Side note, users will see 400 deleted items in the action's logs despite their configuration only allowing 200 and might get confused.

File: src/CleanSoftwareCron.php (hunk 2 above)
Comment (flegastelois) [reply src/CleanSoftwareCron.php:194]: I will suggest another way to fix.

File: src/CleanSoftwareCron.php (hunk 1 above)
Comment (cedric-anne) [reply src/CleanSoftwareCron.php:96]: Maybe handling it in 2 queries does not make sense then. You may delete some `SoftwareVersion` and then delete some `Software` that are not even parents of deleted `SoftwareVersion`.
`getSoftwareWithNoVersionsCriteria()` should probably also return softwares that have none of their version installed, and so the deletion could be made in a unique step.

File: src/CleanSoftwareCron.php [hunk 3]
```diff
@@ -194,17 +194,19 @@ protected static function deleteItems(
         $total = 0;
 
         do {
-            $scope['LIMIT'] = min($max - $total, self::MAX_BATCH_SIZE);
+
+            $scope['LIMIT'] = ($max > self::MAX_BATCH_SIZE) ? self::MAX_BATCH_SIZE : $max;
```
Comment (cedric-anne) [inline src/CleanSoftwareCron.php:198]:
```suggestion
            $scope['LIMIT'] = $max > self::MAX_BATCH_SIZE ? self::MAX_BATCH_SIZE : $max;
```

File: src/CleanSoftwareCron.php [hunk 4]
```diff
@@ -194,17 +194,19 @@ protected static function deleteItems(
...
+            $max -= $count;
 
             foreach ($items as $item) {
                 $em->delete($item);
             }
 
-           // Stop if no items found
-        } while ($count > 0);
+           // Stop if no items found or max is reached
+        } while (($count > 0) && ($max > 0));
```
Comment (cedric-anne) [inline src/CleanSoftwareCron.php:209]:
```suggestion
        } while ($count > 0 && $max > 0);
```

**Review by cedric-anne — COMMENTED** [review]
You should not use so many parenthesis :)

File: src/CleanSoftwareCron.php [hunk 5]
```diff
@@ -191,13 +191,23 @@ protected static function deleteItems(
     ): int {
         global $DB;
 
-        $scope['LIMIT'] = min($max, self::MAX_BATCH_SIZE);
+        $total = 0;
+
+        do {
+
+            $scope['LIMIT'] = ($max > self::MAX_BATCH_SIZE) ? self::MAX_BATCH_SIZE : $max;
```
Comment (AdrienClairembault) [inline src/CleanSoftwareCron.php]: I think it could be simplified to this:
```suggestion
            $scope['LIMIT'] = min($max, self::MAX_BATCH_SIZE)
```

**Comment by cedric-anne** [comment]
Replaced by #14335.
